sdk_contrib/hapi/lib/xray.js

```diff
@@ -0,0 +1,161 @@
+ * @param {Object} [options.logger] Logger to pass to xray
+ * @param {Boolean} [options.automaticMode] Puts xray into automatic or manual mode. Default is true (automatic)
+ * @param {Object[]} [options.plugins] Array of AWS plugins to pass to xray
+ */
+exports.setup = options => {
+  const localOptions = { ...defaultOptions, ...options };
+
+  if (localOptions.logger) {
+    AWSXray.setLogger(localOptions.logger);
+  }
+
+  const segmentName =
+    localOptions.segmentName || exports._internals.createSegmentName();
+  mwUtils.setDefaultName(segmentName);
+
+  if (localOptions.automaticMode) {
+    AWSXray.enableAutomaticMode();
+  } else {
+    AWSXray.enableManualMode();
+  }
+
+  if (localOptions.plugins) {
+    AWSXray.config(localOptions.plugins);
+  }
+
+  if (localOptions.captureAWS) {
+    AWSXray.captureAWS(require('aws-sdk'));
+  }
+
+  if (localOptions.captureHTTP) {
+    AWSXray.captureHTTPsGlobal(require('http'), true);
+    AWSXray.captureHTTPsGlobal(require('https'), true);
+  }
+
+  if (localOptions.capturePromises) {
+    AWSXray.capturePromise();
+  }
+};
+
+/**
+ * Handles the hapi request to set up the XRay segment
+ */
+exports.handleRequest = async (request, h) => {
+  const header = mwUtils.processHeaders(request);
+  const name = mwUtils.resolveName(request.headers.host);
+
+  const segment = new AWSXray.Segment(
+    name,
+    header.Root || header.root,
+    header.Parent || header.parent
+  );
+
+  const { req, res } = request.raw;
+  res.req = req;
+
+  mwUtils.resolveSampling(header, segment, res);
+
+  segment.addIncomingRequestData(new mwUtils.IncomingRequestData(req));
+
+  mwUtils.middlewareLog('Starting Hapi XRay segment', request.url, segment);
+
+  if (AWSXray.isAutomaticMode()) {
+    const ns = AWSXray.getNamespace();
+    if (!request.app) {
+      request.app = {};
+    }
+    request.app.xrayNamespace = ns;
+    request.app.xrayContext = ns.createContext();
+    ns.bindEmitter(req);
+    ns.bindEmitter(res);
+    ns.enter(request.app.xrayContext);
+    AWSXray.setSegment(segment);
+  }
+  request.segment = segment;
+
+  return h.continue;
+};
+
+exports.handleResponse = request => {
+  try {
+    exports._internals.endSegment(request);
+  } finally {
+    if (request.app) {
+      const { xrayNamespace, xrayContext } = request.app;
+
+      if (xrayNamespace && xrayContext) {
+        xrayNamespace.exit(xrayContext);
+      }
+    }
+  }
+};
+
+exports.handleError = (request, error) => {
+  const { segment } = request;
+
+  if (segment) {
+    segment.addError(error);
+    mwUtils.middlewareLog('Hapi XRay segment encountered an error', request.url, segment);
+  }
+};
+
+exports._internals = {
+  endSegment: function(request) {
+    const { segment, response } = request;
+
+    if (!segment || segment.isClosed()) {
+      return;
+    }
+
+    if (response) {
+      if (response.statusCode >= 400) {
+        if (response.statusCode === 429) {
+          segment.addThrottleFlag();
```

Comment:
JFYI I filed #330 since it seems a bit off for each middleware to have this logic. But doesn't affect this PR

Reply:
@anuraaga Sounds good. Happy to tweak this when #330 is addressed.